### PR TITLE
fix beatsync_phase depending on quantize

### PR DESCRIPTION
**Pull request by daschuer**
This PR fixes 
https://bugs.launchpad.net/mixxx/+bug/1772525
and 
https://bugs.launchpad.net/mixxx/+bug/1772526

right click on sync seeking in phase.

### CODE REVIEW

**Comment by daschuer** [comment]
This should also go to 2.1.1. 
Who likes to review?

**Comment by daschuer** [comment]
I really like to have this included in 2.1.1 who can have a look?

**Comment by Be-ing** [comment]
This seems okay from brief testing, but I am wary to introduce such a big UX change in a point release. Are you sure this won't be unwanted in some use cases?

**Comment by daschuer** [comment]
Thank you for review. 
I am pretty sure, because currently the right click in Shade and Deere does the same as the left click. 
After this change, you sync also the phase with the right click, so it is an additional, recently requested feature, for the else redundant right click. 
If you have a different opinion here, I can revert the skin changes and propose them for 2.2. 
What do you think?

**Comment by Be-ing** [comment]
This changes the behavior of `sync_enabled` as well as `beatsync`.

**Comment by rryan** [comment]
No tests?

**Comment by daschuer** [comment]
MM, that's odd, reading this code 
https://github.com/mixxxdj/mixxx/blob/d8fe4dd16e3b69e0ea293c08cc3e8914602018d6/src/engine/sync/enginesync.cpp#L168
It was original implemented that sync_enable also syncs phase.
But I think that was wrong and we have evidentially fixed the "issue"? 

How continue here? 
* In 2.1 right and left click do not sync phase 
* Original left click syncs phase, right click not
* Ieft  click not syncing phase works natural for me. I do not want to have a shift if with quantize off. 
* beatsync should really sync phase and BPM as documented. 

I propose to keep the "buggy" behavior of sync_enable, but keep all other changes of this branch targeting it to 2.1. 

Any other ideas?

**Comment by ywwg** [comment]
> No tests?

agree, the sync code is very brittle and has benefited tremendously from the exhaustive unit tests to keep it from regressing.  It sounds like there should at least be a test that confirms that phase is not synced when sync is pressed.

**Comment by ywwg** [comment]
> But I think that was wrong

I believe on other programs, pushing sync would sync both bpm and phase, especially when both decks were playing.  I replicated that behavior when writing this code. It's worth checking to see what happens when the decks are not synced, and are both playing, and then you click sync.

**Comment by daschuer** [comment]
The linked text was introduced here: https://github.com/mixxxdj/mixxx/pull/133
I have touched and probably brake something here: https://github.com/mixxxdj/mixxx/pull/738
So the current 2.1.0 behaviour was like this since 2.0 without much complains.

From the comments in #133 it looks like coupling the phase sync with quantize is desired.

For me this makes fully sense that this is coupled to quantize. This way the user can enable quantize when there is a good beatgrid and can disable it otherwise. If you enable quantize during sync enabled, the track moves slowly in phase. This matches the initial behaviour as well.

So IMHO the only thing to fix (change) here, is "beatsync" and  "beatsync_phase"
It should match the docs on the wiki page.

Do you agree?

And yes, that should be verified by a test.

**Comment by daschuer** [comment]
OK, now the sync_enabled behaviour is unchanged, beatsync_phase works independently from quantize and a test for both is in place. Is this OK now?

File: src/test/enginesynctest.cpp [hunk 1]
```diff
@@ -1429,3 +1429,52 @@ TEST_F(EngineSyncTest, ZeroBpmNaturalRate) {
     EXPECT_EQ(0.0,
               ControlObject::getControl(ConfigKey(m_sGroup1, "local_bpm"))->get());
 }
+
+TEST_F(EngineSyncTest, SyncPhaseToIfQuantize) {
```
Comment (ywwg) [inline src/test/enginesynctest.cpp]: how about "QuantizeImpliesSyncPhase"  (in your wording, it should be "too")

File: src/test/enginesynctest.cpp [hunk 2]
```diff
@@ -1429,3 +1429,52 @@ TEST_F(EngineSyncTest, ZeroBpmNaturalRate) {
...
+
+    ControlObject::getControl(ConfigKey(m_sGroup1, "play"))->set(1.0);
+    ControlObject::getControl(ConfigKey(m_sGroup2, "play"))->set(1.0);
+    ProcessBuffer();
+
+    // first test without quantisation
+    pButtonSyncEnabled1->set(1.0);
+    ProcessBuffer();
+    // 0.02 where set after testing both cases.
+    EXPECT_LT(0.02, ControlObject::getControl(ConfigKey(m_sGroup1, "beat_distance"))->get());
```
Comment (ywwg) [inline src/test/enginesynctest.cpp]: expect less-than?  Should this be expect_float_eq or similar instead?

File: res/skins/Shade/style.qss [hunk 3]
```diff
@@ -83,32 +83,202 @@ WEffectSelector::checked
...
+	#LibraryBPMButton::item:checked:hover {
+	   border: 1px solid #898989;
+	}
+
+  #LibraryBPMSpinBox {
+    color: #fff;
+    selection-color: #0f0f0f;
+    background-color: #0f0f0f;
+    selection-background-color: #888;
+    /* This combination of negative top/bottom padding & margin
```
Comment (ywwg) [inline res/skins/Shade/style.qss]: I think your branch is behind a bit, causing these library CSS diffs

**Review by ywwg — COMMENTED** [review]
my only real question is about the use of EXPECT_LT in the tests, otherwise this looks fine.  thanks!

**Comment by Be-ing** [comment]
> So IMHO the only thing to fix (change) here, is "beatsync" and "beatsync_phase"

This would still change the behavior of the [SyncButton in Components](https://github.com/mixxxdj/mixxx/blob/master/res/controllers/midi-components-0.0.js#L245).

**Comment by ywwg** [comment]
My recollection is that "beatsync" was kept around for backwards-compatibility purposes.  A short-press of sync_enable is indeed identical.

**Comment by Be-ing** [comment]
> A short-press of sync_enable is indeed identical.

Not when set through a controller script.

**Comment by daschuer** [comment]
@Be-ing can you explain why we need a hack in Components? Why can't  you use plain sync_enable, to get the desired behaviour?

Do I understand it right, you wish that we keep bratsync quantize dependent?
If yes, we might  need a unconditional alternative in 2.2 as requested, to make Mixxx behave like other tools. I am happy without it though. "Beatsync_tempo_phase" ?

**Comment by Be-ing** [comment]
> @Be-ing can you explain why we need a hack in Components?

The push-and-hold behavior for sync_enabled was implemented as a hack that only affects XML mappings and skins, not controller scripts.

**Comment by Be-ing** [comment]
> The push-and-hold behavior for sync_enabled was implemented as a hack that only affects XML mappings and skins, not controller scripts.

IMO this should not be changed because there should be a way to immediately activate master sync without having to push and hold a button.

**Comment by ywwg** [comment]
what's wrong with push and hold?

**Comment by Be-ing** [comment]
It's easy to forget to hold, which can easily mess up a mix.

**Comment by ywwg** [comment]
has this happened or are you speculating?  on my controller sync is a lighted button, so when I push and hold it it stays lit.  It's easy to tell when master sync is on or off based on the lighted state of the button, so there's nothing to forget.

**Comment by Be-ing** [comment]
Yes that happened to me when I had my controller mapped with push-and-hold, so I didn't keep it mapped that way very long and switched back to a simple toggle button. Seeing the master sync state on the controller doesn't solve the problem of having to hold down a button and manipulate other controls for a mix within a short span of time.

**Comment by ywwg** [comment]
I'm a little unclear of the use-case you describe.  When I use master sync I turn it on at the beginning of my set and it stays on the whole time.  Or if I'm turning it on it's during the previewing stage where I'm not in a rush.  What were you doing that you needed to turn it on quickly while you were doing other things?

**Comment by Be-ing** [comment]
I only use sync occasionally, so I have it off most of the time.

**Comment by ywwg** [comment]
is push-and-hold an operation on the buttons, or on the control object? Maybe it should be up to the controller configuration to decide how that works.

**Comment by daschuer** [comment]
It is an option of the ControlObject. Ok so i will restore the beatsync behavior.

**Comment by Be-ing** [comment]
> is push-and-hold an operation on the buttons, or on the control object? Maybe it should be up to the controller configuration to decide how that works.

I don't understand what you are asking. What do you mean by "an operation on the buttons"? An operation on the Button Component in JavaScript?

> you cannot give a button release command from Controller scripts

Yes you can by setting the ControlObject to 0 like other ControlPushButtons. However, the `sync_enabled` ControlObject couples the value of the ControlObject to an indicator of whether master sync is enabled, so it cannot be used this way. Perhaps a way out of this situation would be to make a new ControlObject for sync button input and keep `sync_enabled` for indicating the state of master sync. This would be similar to how we have separate Control objects for `play`/`play_indicator` and `cue_default`/`cue_indicator`. Then the new ControlObject could either enable master sync on short press or long press depending on a new checkbox in the Decks section of the preferences. I propose the name `sync` for this new ControlObject.

**Comment by daschuer** [comment]
No, it is a different situation here. 

You can't fake a button release by setting the value to zero. If you set the control object to 0 it gos immediately to 0 bypassing a toggle latching and power window code.  
We need to expose: 
ControlObject::setValueFromMidi(MidiOpCode o, double v)
in order to do this form scripts. 
Maybe we should wrap it into a press() and release() function and we are done. I bet we can delete a lot of Components hacks using this.

**Comment by daschuer** [comment]
You can, just set the value to 1.0 and it is permanent enabled. Or call upcoming push().
If you call the push() and release() functions in a row,  you have beatsync behaviour.

**Comment by daschuer** [comment]
OK, done. Wiki and tooltips updated as well.

**Comment by Be-ing** [comment]
This has a lot of unrelated changes in it now so it is hard to review. Please rebase it on the target branch.

**Comment by daschuer** [comment]
Ups, I have messed up my 2.1 branch. Fixed now.

**Comment by daschuer** [comment]
@Be-ing:ready to merge? I would like to release 2.1.1 including this.

**Comment by daschuer** [comment]
This PR does not introduce any new control.
I think you mean the existing control beatsync_phase. 
Before this PR it does nothing in case of quantize of. Now it always independent from quantize and seeks the track on which the button is pressed. No idea why it is the other track in your case.

**Comment by daschuer** [comment]
I have double checked it and only the track where I press sync does the seek. 
Do you have a recipe to reproduce it?

**Comment by daschuer** [comment]
@Be-ing: merge?

File: src/engine/sync/synccontrol.cpp [hunk 4]
```diff
@@ -161,12 +163,18 @@ void SyncControl::notifyOnlyPlayingSyncable() {
     m_pBpmControl->resetSyncAdjustment();
 }
 
-void SyncControl::requestSyncPhase() {
-    m_pChannel->getEngineBuffer()->requestSyncPhase();
+void SyncControl::requestSync() {
+    if (isPlaying() && m_pQuantize->toBool()) {
+        // only sync phase if the deck is playing and if quantize is enabled.
+        // this way the it is up to the user to decide if a seek is desired or not.
+        // This is helpfull if the beatgrid of the track doe not fit at the current
```
Comment (Be-ing) [inline src/engine/sync/synccontrol.cpp]: helpful

File: src/engine/sync/synccontrol.cpp [hunk 5]
```diff
@@ -161,12 +163,18 @@ void SyncControl::notifyOnlyPlayingSyncable() {
...
+        // payposition
```
Comment (Be-ing) [inline src/engine/sync/synccontrol.cpp]: playposition

File: src/test/enginesynctest.cpp (hunk 1 above)
Comment (Be-ing) [reply src/test/enginesynctest.cpp]: ping

**Comment by Be-ing** [comment]
> I'm confused what this new sync_phase Control is intended to do. Now right clicking the sync button makes the track seek so the play position jumps to the beat grid of the track that sync_phase was pressed on. Isn't it meant to align the beatgrid of the track sync_phase is pressed on to the other track's beatgrid?

I think I got confused before because I tested tracks with very different tempos. `sync_phase` works as described.

File: src/test/enginesynctest.cpp [hunk 6]
```diff
@@ -1429,3 +1429,73 @@ TEST_F(EngineSyncTest, ZeroBpmNaturalRate) {
...
+    pButtonBeatsync1->set(1.0);
+    ProcessBuffer();
+
+    // 0.19933910991038406 in case quantize is disabled
+    ASSERT_DOUBLE_EQ(0.19350798541791794, ControlObject::get(ConfigKey(m_sGroup1, "beat_distance")));
+
+}
+
+
+
```
Comment (Be-ing) [inline src/test/enginesynctest.cpp]: extra newlines at end of file

File: src/engine/enginebuffer.cpp [hunk 7]
```diff
@@ -1179,7 +1179,7 @@ void EngineBuffer::processSeek(bool paused) {
             return;
     }
 
-    if ((seekType & SEEK_PHASE) && !paused && m_pQuantize->toBool()) {
+    if (!paused && ((seekType & SEEK_PHASE) || m_pQuantize->toBool())) {
```
Comment (Be-ing) [inline src/engine/enginebuffer.cpp]: I am not familiar with this area of the code. Can you clarify what this change does?

**Review by Be-ing — CHANGES_REQUESTED** [review]
I am a little confused now what this PR changes at this point. IIUC the only behavior change is that `beatsync` only syncs phase if quantize is enabled now?

**Comment by daschuer** [comment]
> I am a little confused now what this PR changes at this point. IIUC the only behavior change is that beatsync only syncs phase if quantize is enabled now?

The changes are that beatsync_phase syncs always phase even if quantize is disabled and we put that to the right klick instead of other redundant controls.
The other controls have not chsnges, and the description in the Mixxx Wikki have been corrected.

File: src/engine/enginebuffer.cpp (hunk 7 above)
Comment (daschuer) [reply src/engine/enginebuffer.cpp]: This allows to sync phase independent from quantize now. The sync signal of the controls that must not sync phase if quatize is of, is done in their buissnes logic.

Comment (Be-ing) [reply src/engine/enginebuffer.cpp]: Okay thank you for explaining.

**Comment by daschuer** [comment]
Done!

**Comment by Be-ing** [comment]
I'm still confused. I just tested master and `beatsync_phase` works regardless of whether quantize is enabled and `beatsync` only syncs phase if quantize is enabled. So what does this change other than the skins?

**Comment by daschuer** [comment]
beatsync_phase did not work independently from quantize before.

**Comment by Be-ing** [comment]
> I just tested master and beatsync_phase works regardless of whether quantize is enabled

I checked again and couldn't reproduce that. So this PR LGTM.

**Comment by daschuer** [comment]
Thank you.

**Pull request merged**